File: forge/solve_camera.py
#!/usr/bin/env python3
"""Solve a camera (x, z, heading, hfov) from landmark horizontal positions in a video frame.

usage: solve_camera.py landmarks.json
landmarks.json: {"landmarks": [{"name": "corner", "x": -3, "z": -3, "frac": 0.55}, ...],
                 "aspect": 1.778, "y": 1.05, "search": {"x": [0, 8], "z": [0, 8]}}
frac = horizontal position in the frame (0 = left edge, 1 = right edge).
Prints CAM, LOOK (6 m ahead at camera height) and the vertical fov for three.js.
"""
import json, math, sys

def main(path):
    cfg = json.load(open(path))
    L = cfg["landmarks"]; aspect = cfg.get("aspect", 16 / 9); cy = cfg.get("y", 1.05)
    sx = cfg.get("search", {}).get("x", [-2, 10]); sz = cfg.get("search", {}).get("z", [-2, 10])
    best = None
    for hfov in [60, 65, 70, 75, 80, 85, 90, 95]:
        t = math.tan(math.radians(hfov / 2))
        for i in range(0, 41):
            cx = sx[0] + (sx[1] - sx[0]) * i / 40
            for j in range(0, 41):
                cz = sz[0] + (sz[1] - sz[0]) * j / 40
                angs = [math.atan2(l["z"] - cz, l["x"] - cx) for l in L]
                # heading = mean angle weighted to center the landmarks
                for h_i in range(0, 72):
                    h = -math.pi + h_i * math.pi / 36
                    err = 0.0
                    for a, l in zip(angs, L):
                        d = (a - h + math.pi) % (2 * math.pi) - math.pi
                        pred = 0.5 + math.tan(d) / (2 * t) if abs(d) < math.pi / 2 else 9
                        err += (pred - l["frac"]) ** 2
                    if best is None or err < best[0]:
                        best = (err, cx, cz, h, hfov)
    err, cx, cz, h, hfov = best
    vfov = 2 * math.degrees(math.atan(math.tan(math.radians(hfov / 2)) / aspect))
    lx, lz = cx + 6 * math.cos(h), cz + 6 * math.sin(h)
    print(f"rms error {math.sqrt(err / len(L)):.3f} (fraction of frame width)")
    print(f"CAM  = ({cx:.2f}, {cy:.2f}, {cz:.2f})   heading {math.degrees(h):.1f} deg   hfov {hfov} -> three.js fov {vfov:.1f}")
    print(f"LOOK = ({lx:.2f}, {cy:.2f}, {lz:.2f})")
    for a, l in zip([math.atan2(l["z"] - cz, l["x"] - cx) for l in L], L):
        d = (a - h + math.pi) % (2 * math.pi) - math.pi
        pred = 0.5 + math.tan(d) / (2 * math.tan(math.radians(hfov / 2)))
        print(f"  {l['name']:<14} observed {l['frac']:.2f}  predicted {pred:.2f}")

if __name__ == "__main__":
    main(sys.argv[1] if len(sys.argv) > 1 else "landmarks.json")

File: forge/solve_camera.mjs
#!/usr/bin/env node
// solve_camera.mjs — solve a camera from measurements taken off a reference frame.
//
// Give it two landmarks whose world positions you know, the NDC position each
// one occupies in the frame, and (for at least one) its apparent size. It grid-
// searches camera position, heading and fov, refining four times, and prints a
// CAM / LOOK / fov you can paste into a shot.
//
//   node solve_camera.mjs '{
//     "aspect": 2.34,
//     "A": {"p":[0,0,0],      "r":11, "ndc":[0.45,0.15], "sizeNdc":0.72},
//     "B": {"p":[-46,-12,10], "r":15, "ndc":[-0.47,-0.22], "sizeNdc":0.66},
//     "range": {"fov":[26,56,2], "d":[36,150,3], "az":[-175,175,5], "el":[-50,50,4]}
//   }'
//
// ndc: x right, y up, ±1 at the frame edge.
// sizeNdc: the body's on-screen RADIUS as a fraction of the frame's HALF-height.
//   Measure a radius in pixels, divide by (frameHeightPx / 2). For a width
//   measurement multiply by the aspect ratio first.
//
// The camera is aimed so that A lands exactly on its NDC; B's NDC and both
// sizes are what the search minimises. Constrain `el` when the solution family
// is ambiguous — it usually is, and the wrong branch puts the camera above a
// subject the reference clearly looks up at.
const S = JSON.parse(process.argv[2]);
const A = S.aspect ?? 1972 / 842;
const sub = (a, b) => [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
const dot = (a, b) => a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
const cross = (a, b) => [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]];
const len = (a) => Math.sqrt(dot(a, a));
const nrm = (a) => { const l = len(a); return [a[0] / l, a[1] / l, a[2] / l]; };
const add = (a, b) => [a[0] + b[0], a[1] + b[1], a[2] + b[2]];
const mul = (a, k) => [a[0] * k, a[1] * k, a[2] * k];
const basis = (f) => { const r = nrm(cross(f, [0, 1, 0])); return [r, cross(r, f)]; };

function proj(cam, f, P, fov) {
  const ty = Math.tan(fov * Math.PI / 360), tx = ty * A, [r, u] = basis(f);
  const d = sub(P, cam), z = dot(d, f);
  if (z <= 0) return null;
  return [dot(d, r) / z / tx, dot(d, u) / z / ty];
}
// rotate the view direction until `P` lands on (nx, ny)
function aim(cam, P, nx, ny, fov) {
  const ty = Math.tan(fov * Math.PI / 360), tx = ty * A, dn = nrm(sub(P, cam));
  let f = dn.slice();
  for (let i = 0; i < 60; i++) {
    const [r, u] = basis(f);
    const want = nrm(add(add(f, mul(r, nx * tx)), mul(u, ny * ty)));
    const v = cross(want, dn), c = dot(want, dn);
    if (len(v) < 1e-12) break;
    const k = 1 / (1 + c);
    const rot = (x) => add(add(x, cross(v, x)), mul(cross(v, cross(v, x)), k));
    f = nrm(rot(f));
  }
  return f;
}

let R = S.range || { fov: [26, 60, 2], d: [30, 190, 4], az: [-175, 175, 5], el: [-70, 40, 4] };
let best = null;
function search(R) {
  let b = null;
  for (let fov = R.fov[0]; fov <= R.fov[1]; fov += R.fov[2])
  for (let d = R.d[0]; d <= R.d[1]; d += R.d[2])
  for (let az = R.az[0]; az <= R.az[1]; az += R.az[2])
  for (let el = R.el[0]; el <= R.el[1]; el += R.el[2]) {
    const a = az * Math.PI / 180, e = el * Math.PI / 180;
    const cam = [S.A.p[0] + d * Math.cos(e) * Math.sin(a),
                 S.A.p[1] + d * Math.sin(e),
                 S.A.p[2] + d * Math.cos(e) * Math.cos(a)];
    const f = aim(cam, S.A.p, S.A.ndc[0], S.A.ndc[1], fov);
    const q = proj(cam, f, S.B.p, fov);
    if (!q) continue;
    const ty = Math.tan(fov * Math.PI / 360);
    let err = Math.hypot(q[0] - S.B.ndc[0], q[1] - S.B.ndc[1]);
    if (S.A.sizeNdc) err += Math.abs((S.A.r / d) / ty - S.A.sizeNdc) * 2;
    if (S.B.sizeNdc) {
      const dB = len(sub(S.B.p, cam));
      err += Math.abs((S.B.r / dB) / ty - S.B.sizeNdc) * 2;
    }
    if (!b || err < b.err) b = {
      err: +err.toFixed(4), fov: +fov.toFixed(2), d: +d.toFixed(2), az, el,
      cam: cam.map((v) => +v.toFixed(2)),
      look: add(cam, mul(f, d)).map((v) => +v.toFixed(2)),
      Bndc: q.map((v) => +v.toFixed(3)),
    };
  }
  return b;
}
for (let pass = 0; pass < 4; pass++) {
  best = search(R);
  if (!best) { console.error('no solution in range'); process.exit(1); }
  const z = (k, st) => [best[k] - st * 2, best[k] + st * 2, st / 2.5];
  R = { fov: z('fov', R.fov[2]), d: z('d', R.d[2]), az: z('az', R.az[2]), el: z('el', R.el[2]) };
}
console.log(JSON.stringify(best, null, 1));
console.log(`cam ${best.cam}  look ${best.look}  fov ${best.fov}`);
